upup/pkg/fi/cloudup/gce/tpm/gcetpmverifier/tpmverifier.go

```diff
@@ -0,0 +1,247 @@
+func NewTPMVerifier(opt *gcetpm.TPMVerifierOptions) (bootstrap.Verifier, error) {
+	ctx := context.Background()
+
+	computeClient, err := compute.NewService(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error building compute API client: %w", err)
+	}
+
+	return &tpmVerifier{
+		opt:           *opt,
+		computeClient: computeClient,
+	}, nil
+}
+
+var _ bootstrap.Verifier = &tpmVerifier{}
+
+func (v *tpmVerifier) VerifyToken(ctx context.Context, authToken string, body []byte) (*bootstrap.VerifyResult, error) {
+	if !strings.HasPrefix(authToken, gcetpm.GCETPMAuthenticationTokenPrefix) {
+		return nil, fmt.Errorf("incorrect authorization type")
+	}
+	authToken = strings.TrimPrefix(authToken, gcetpm.GCETPMAuthenticationTokenPrefix)
+
+	tokenBytes, err := base64.StdEncoding.DecodeString(authToken)
+	if err != nil {
+		return nil, fmt.Errorf("decoding authorization token: %w", err)
+	}
+
+	token := &gcetpm.AuthToken{}
+	err = json.Unmarshal(tokenBytes, token)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling authorization token: %w", err)
+	}
+
+	// Guard against replay attacks
+	if token.Data.Audience != gcetpm.AudienceNodeAuthentication {
+		return nil, fmt.Errorf("incorrect Audience")
+	}
+	timeSkew := math.Abs(time.Since(time.Unix(token.Data.Timestamp, 0)).Seconds())
+	if timeSkew > float64(v.opt.MaxTimeSkew) {
+		return nil, fmt.Errorf("incorrect Timestamp %v", token.Data.Timestamp)
+	}
+
+	// Verify the token has signed the body content.
+	requestHash := sha256.Sum256(body)
+	if !bytes.Equal(requestHash[:], token.Data.RequestHash) {
+		return nil, fmt.Errorf("incorrect RequestHash")
+	}
+
+	// Some basic validation to avoid requesting invalid instances.
+	if token.Data.GCPProjectID == "" {
+		return nil, fmt.Errorf("gcpProjectID is required")
+	}
+	if token.Data.Zone == "" {
+		return nil, fmt.Errorf("zone is required")
+	}
+	if token.Data.Instance == "" {
+		return nil, fmt.Errorf("instance is required")
+	}
+
+	// Verify node is in our cluster
+	if token.Data.GCPProjectID != v.opt.ProjectID {
+		return nil, fmt.Errorf("projectID does not match expected: got %q, want %q", token.Data.GCPProjectID, v.opt.ProjectID)
+	}
+
+	instance, err := v.computeClient.Instances.Get(token.Data.GCPProjectID, token.Data.Zone, token.Data.Instance).Do()
+	if err != nil {
+		if isNotFound(err) {
+			return nil, fmt.Errorf("unable to find instance in compute API: %w", err)
+		}
+		return nil, fmt.Errorf("error fetching instance from compute API: %w", err)
+	}
+
+	if !strings.HasPrefix(lastComponent(instance.Zone), v.opt.Region+"-") {
+		return nil, fmt.Errorf("instance was in zone %q, expected region %q", instance.Zone, v.opt.Region)
+	}
+
+	clusterName := ""
+	instanceGroupName := ""
+	for _, item := range instance.Metadata.Items {
+		switch item.Key {
+		case gce.MetadataKeyInstanceGroupName:
+			instanceGroupName = fi.StringValue(item.Value)
+		case "cluster-name":
+			clusterName = fi.StringValue(item.Value)
```

Comment:
One thing to consider, perhaps not for this PR, is how to allow an instance to revoke its privilege to make this query of kops-controller. That way, any malicious code later obtaining access to the TPM wouldn't be able to escalate to having kubelet access. For AWS, my employer has a hook run after nodeup, sending a request to a lambda that changes the instance role to one with minimal privileges.

I think for GCE you could have it add or remove an instance metadata item, which this code would check.

Reply:
That's a really good idea - though I agree that as it isn't something we can do in the TPM itself (I assume!) it probably doesn't belong in this PR.  We could also consider having kops-controller remember that it issued the cert (maybe mark the node?) / refuse to issue to a machine that had been up for more than N minutes.